### DISCUSSION TITLE
Finite differences Jacobian

### DISCUSSION THREAD

**Discussion by ConnorStoneAstro** [Autostronomy/AstroPhot#71] [Ideas]
Automatic differentiation is nice for getting exact derivatives, however it is very memory intensive for large images with many models. Finite differences uses considerably less memory and so should be included.

**Comment by ranbir7** [comment]
how shall I approach this?
Im really interested in participating

**Comment by ConnorStoneAstro** [comment]
This is perhaps one of the most complex issues in the list (if you're looking for a challenge then issues 4, 6, and 8 would all be good options). Essentially what is needed is that you iterate through each parameter (first construct the full flattened parameter tensor then go one element at a time) and add a small epsilon change to it's value (say 1e-4), then re-run the model.full_sample method and take a difference between the original image and the re-run with param+epsilon. That difference divided by epsilon should be the jacobian for that parameter. Do it for each parameter and store the results in a format like the other jacobian functions.

**Comment by ConnorStoneAstro** [comment]
Thinking about it further, it is perhaps not as difficult as I originally envisioned. The place to fill the code is in the  ``jacobian_finite`` function of the ``autoprof/models/model_object.py`` module. To start, you can call the ``self.get_parameter_vector`` function, this will return a 1D tensor. Then you iterate through each element of the tensor and add epsilon (described above, probably 1e-4, but make it a class variable), then call the ``self.full_sample`` function with the new parameter vector. Compare the new sample with a sample done at the original vector (so: (new_sample - original_sample)/epsilon) and that is the jacobian for one parameter. Then do this for each parameter and fill the jacobian tensor which should have a shape just like in the jacobian function, which I think is (*image shape, N parameters).

**Comment by ranbir7** [comment]
okay okay 
lemme try understanding it
PLUS another thing 
i get error when I run autoprof command in terminal 
I have already installed it 
![image](https://user-images.githubusercontent.com/71406920/217358774-29817779-a70f-44df-80c7-90b365489982.png)

**Comment by ranbir7** [comment]
I got it 
the command needed some extra arguments and It worked perfectly 
when I ran "autoprof tutorials" 
I'll be looking into Jacobian Issue after I get to understand whats happening.

**Comment by ConnorStoneAstro** [comment]
Let me know if anything is unclear about the jacobian. As for running autoprof in the terminal, you have to give it a config file for it to run. See the getting started tutorial: https://connorstoneastro.github.io/AutoProf-2/getting_started.html
and the command line section is at: https://connorstoneastro.github.io/AutoProf-2/configfile_interface.html

ah, I see you have it working, nice! let me know if you run into any issues. For sure I will update the code to give a more helpful error message so future users don't get confused by the AttributeError above!

**Comment by ranbir7** [comment]
hey 
what does astropy.io module do in the simple config file of docs/tutorials ?
I cant seem to find it in the internet.

**Comment by ConnorStoneAstro** [comment]
Ah, astropy is a package with lots of astronomy related tools. In this case it is being used to write a `.fits` image file which is the standard format in astronomy.

**Comment by ConnorStoneAstro** [comment]
May be possible to avoind finitte differences jacobian by creating a system to "chunk" the jacobian calculation into smaller sub-windows of the total image.

**Comment by ConnorStoneAstro** [comment]
After various recent optimizations to the autograd jacobian methods, the use of finite differences jacobians no longer seems so critical. For now I will close this discussion, but we can re-open it if there ends up being a special use-case.
